### PR TITLE
[core] Make case insensitive dict public

### PR DESCRIPTION
**Pull request by iscai-msft**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by xiangyan99 — COMMENTED** [review]
1.23.0.

And changelog?

**Comment by iscai-msft** [comment]
@xiangyan99 thanks, let me add now

File: sdk/core/azure-core/azure/core/utils/_utils.py [hunk 1]
```diff
@@ -80,15 +80,14 @@ def _convert_to_isoformat(date_time):
...
-def _case_insensitive_dict(*args, **kwargs):
-    """Return a case-insensitive dict from a structure that a dict would have accepted.
-
-    Rational is I don't want to re-implement this, but I don't want
-    to assume "requests" or "aiohttp" are installed either.
-    So I use the one from "requests" or the one from "aiohttp" ("multidict")
-    If one day this library is used in an HTTP context without "requests" nor "aiohttp" installed,
-    we can add "multidict" as a dependency or re-implement our own.
-    """
+def case_insensitive_dict(*args, **kwargs):
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/utils/_utils.py:83]: Can you add MutableMapping typing?

**Review by lmazuel — APPROVED** [review]

File: sdk/core/azure-core/azure/core/utils/_utils.py [hunk 2]
```diff
@@ -80,15 +81,18 @@ def _convert_to_isoformat(date_time):
...
-    Rational is I don't want to re-implement this, but I don't want
-    to assume "requests" or "aiohttp" are installed either.
-    So I use the one from "requests" or the one from "aiohttp" ("multidict")
-    If one day this library is used in an HTTP context without "requests" nor "aiohttp" installed,
-    we can add "multidict" as a dependency or re-implement our own.
+    :return: A case-insensitive mutable mapping object.
+    :rtype: ~collections.abc.MutableMapping
     """
+
+    # Rational is I don't want to re-implement this, but I don't want
```
Comment (iscai-msft) [inline sdk/core/azure-core/azure/core/utils/_utils.py:91]: when we switch over to our own implementation, my thinking was keeping the implementation private in this file, then having the public method `case_insensitive_dict` return our private implementation instead

**Review by xiangyan99 — APPROVED** [review]

**Review by annatisch — APPROVED** [review]

**Pull request merged**